src/relay/libp2p_relay_server.erl

```diff
@@ -105,73 +115,73 @@ handle_cast(init_relay, #state{started=false, swarm=Swarm}=State0) ->
-handle_info(try_relay, State = #state{started=false}) ->
-    case int_relay(State) of
-        {ok, Connection} ->
-            lager:debug("relay started successfuly"),
-            {noreply, add_flap(State#state{started=true, connection=Connection})};
-        {error, no_peer} ->
-            lager:warning("could not initiate relay no peer found"),
-            {noreply, State};
-        {error, no_address} ->
-            lager:warning("could not initiate relay, swarm has no listen address"),
-            {noreply, State};
+handle_info({new_peers, NewPeers}, #state{tid=TID, stream=Pid}=State) when is_pid(Pid) ->
+    Swarm = libp2p_swarm:swarm(TID),
+    SwarmAddr = libp2p_swarm:pubkey_bin(Swarm),
+    {noreply, State#state{peers=sort_peers(merge_peers(NewPeers, State#state.peers), SwarmAddr), peer_index=1}};
+handle_info({new_peers, NewPeers}, #state{tid=TID}=State) ->
+    Swarm = libp2p_swarm:swarm(TID),
+    SwarmAddr = libp2p_swarm:pubkey_bin(Swarm),
+    {noreply, State#state{peers=sort_peers(merge_peers(NewPeers, State#state.peers), SwarmAddr)}};
+handle_info(retry, #state{stream=undefined}=State) ->
+    case init_relay(State) of
+        {ok, Pid} ->
+            _ = erlang:monitor(process, Pid),
+            lager:info("relay started successfuly with ~p", [Pid]),
+            {noreply, add_flap(State#state{stream=Pid, address=undefined, retrying=undefined})};
         _Error ->
             lager:warning("could not initiate relay ~p", [_Error]),
-            % TODO: exponential back off here?
-            erlang:send_after(2500, self(), try_relay),
-            {noreply, next_peer(State)}
+            {noreply, next_peer(retry(State))}
     end;
+handle_info({'DOWN', _Ref, process, Pid, _Reason}, #state{tid=TID, stream=Pid, address=Address}=State) ->
+    _ = libp2p_config:remove_listener(TID, Address),
+    {noreply, retry(State#state{stream=undefined, address=undefined})};
 handle_info(_Msg, State) ->
-    lager:debug("rcvd unknown info msg: ~p", [_Msg]),
+    lager:warning("rcvd unknown info msg: ~p", [_Msg]),
     {noreply, State}.
 
 code_change(_OldVsn, State, _Extra) ->
     {ok, State}.
 
-terminate(_Reason, #state{tid=TID, connection=Connection}=_State) ->
-    case Connection of
-        undefined -> ok;
-        Pid ->
-            %% stop any active relay connection so that the listen address is removed from the peerbook
-            catch libp2p_framed_stream:close(Pid)
-    end,
+terminate(_Reason, #state{tid=TID, stream=Pid}) when is_pid(Pid) ->
+    catch libp2p_framed_stream:close(Pid),
     true = libp2p_config:remove_relay(TID),
+    ok;
+terminate(_Reason, _State) ->
     ok.
 
 %% ------------------------------------------------------------------
 %% Internal Function Definitions
 %% ------------------------------------------------------------------
+
+retry(#state{retrying=undefined}=State) ->
+    TimeRef = erlang:send_after(2500, self(), retry),
```

Comment:
Should we use the `backoff` library here like we do in group_worker to back of retrying too regularly?

Reply:
we use to have an issue with this but it should be fixed now so I don't think backoff really matters

Reply:
so if we don't have an issue do we still want to retry?

Reply:
the issue was we were retrying all the time sending A LOT of messages. I don't really want to back off here as I want to get an address asap